src/formik.tsx

```diff
@@ -382,34 +386,46 @@ Formik cannot determine which value to update. See docs for more information: ht
 
       handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
         e.preventDefault();
+        this.submitForm();
+      };
+
+      submitForm = () => {
         this.setState({
           touched: touchAllFields(this.state.values),
           isSubmitting: true,
         });
 
         if (validate) {
-          const errors = validate(this.state.values);
-          this.setState({ errors, isSubmitting: errors.keys.length > 0 });
-
-          // only call if there are no errors errors
-          if (errors.keys.length === 0) {
-            this.submitForm();
+          const maybePromisedErrors =
+            validate(this.state.values, this.props) || {};
+          if (isPromise(maybePromisedErrors)) {
+            (validate(this.state.values, this.props) as Promise<any>).then(
```

Comment:
Second call to (validate(values, this.props) as Promise<any>) causes second validation run here. It feasible to run maybePromisedErrors.then() here - not sure whether you loose some of typehints here...